tests/text_renderer_target_specific.rs

```diff
@@ -0,0 +1,119 @@
+use embedded_graphics::{mock_display::MockDisplay, pixelcolor::Rgb888, prelude::*, primitives::Rectangle, text::{Alignment, Baseline, Text, TextStyleBuilder, renderer::{CharacterStyle, TargetSpecificTextRenderer, TextMetrics}}};
+
+type Error = <MockDisplay<Rgb888> as DrawTarget>::Error;
+
+struct TestDisplay(MockDisplay<Rgb888>);
+
+impl DrawTarget for TestDisplay {
+    type Color = Rgb888;
+    type Error = Error;
+
+    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        self.0.draw_iter(pixels)
+    }
+}
+
+impl OriginDimensions for TestDisplay {
+    fn size(&self) -> Size {
+        Size::new_equal(128)
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct DummyTargetTextStyle<T>(T);
+
+impl<T: PixelColor> CharacterStyle for DummyTargetTextStyle<T> {
+    fn measure_string(&self, text: &str, position: Point, _baseline: Baseline) -> TextMetrics {
+        // The `baseline` is ignored in this test, but must be used in a real impl.
+
+        let width = text.len() as u32 * 4;
+
+        TextMetrics {
+            bounding_box: Rectangle::new(position, Size::new(width, 4)),
+            next_position: position + Size::new(width, 0),
+        }
+    }
+
+    fn line_height(&self) -> u32 {
+        5
+    }
+}
+
+impl TargetSpecificTextRenderer<TestDisplay> for DummyTargetTextStyle<Rgb888> {
+    fn draw_string(
+        &self,
+        text: &str,
+        position: Point,
+        baseline: Baseline,
+        target: &mut TestDisplay,
+    ) -> Result<Point, Error> {
+        let metrics = self.measure_string(text, position, baseline);
+
+        target.fill_solid(&metrics.bounding_box, self.0)?;
+
+        Ok(metrics.next_position)
+    }
+
+    fn draw_whitespace(
+        &self,
+        _width: u32,
+        _position: Point,
+        _baseline: Baseline,
+        _target: &mut TestDisplay,
+    ) -> Result<Point, Error> {
+        todo!()
+    }
+}
+
+#[test]
+fn target_specific_text_renderer() {
+    let mut target = TestDisplay(MockDisplay::new());
+
+    Text::new("ab\nc", Point::zero())
+        .into_styled(DummyTargetTextStyle(Rgb888::GREEN))
+        .draw(&mut target)
+        .unwrap();
+
+    target.0.assert_pattern(&[
+        "GGGGGGGG", //
+        "GGGGGGGG", //
+        "GGGGGGGG", //
+        "GGGGGGGG", //
+        "        ", //
+        "GGGG    ", //
+        "GGGG    ", //
+        "GGGG    ", //
+        "GGGG    ", //
+    ]);
+}
+
+#[test]
+fn target_specific_text_renderer_with_text_style() {
+    let mut target = TestDisplay(MockDisplay::new());
+
+    let character_style = DummyTargetTextStyle(Rgb888::BLUE);
+    let text_style = TextStyleBuilder::new()
+        .character_style(character_style)
+        .alignment(Alignment::Right)
+        .build();
```

Comment:
For my use case, this is the biggest limitation of the current approach. In order to use target-specific font rendering techniques, you need to use the target-specific `DummyTargetTextStyle`. This makes the definition of a UI non-portable across e-g- compatible backends. For instance, if I wanted to define a UI library that was display-independent, I would have no way of allowing a user of that library to provide an accelerated font rendering method if one is available.

Put another way, it seems like there is no way to use display-specific font rendering methods if the display is not known ahead of time. I might be wrong though-- I am still getting up to speed on the font system in e-g.

Is there any way the text rendering functionality could be defined as a part of the `DrawTarget` implementation? If it could be defined as part of the `DrawTarget` implementatoin, then a display implementer could define their font rendering method (or not, and therefore fall back to the default) and a library would not have to know any details.

I am aware my suggestion of putting functionality back into `DrawTarget` is in direct tension with the push to separate `DrawTarget` out from dependency on e-g primitives, so I understand if that's not a direction y'all would like to take.

Reply:
If I understand the RA8875 the controller supports rendering of fonts from an internal or external ROM. How could this accelerate the drawing of `MonoFont`s, which aren't included in the ROM?

> Put another way, it seems like there is no way to use display-specific font rendering methods if the display is not known ahead of time. I might be wrong though-- I am still getting up to speed on the font system in e-g.

An UI library could be made generic, so that the library user can either use a generic font or a target specific one. This should provide a way to implement both variants with some conditional compilation in the application. But there might be additional problems I haven't thought about yet.

> Is there any way the text rendering functionality could be defined as a part of the `DrawTarget` implementation?

I don't think there is, if we want to continue to support external text renderers. An external text renderer can use very different draw calls than the internal `MonoFont` text renderer. The `eg-seven-segment` renderer, for example, draws all glyphs using rectangles.